documents filtering (#285)
meilisearch/documentation: guides/advanced_guides/filtering.md

# Filtering

A search filter is a parameter that can be applied to the search query to refine the search results of a particular field (e.g. by size, color, price, or brand).

> Multiple filters may be applied to take a broad range of documents and refine them into a more narrow selection, allowing the end user to retrieve the most relevant search results based on the criteria they've selected.

Reviewer: I'm not sure about what I said but this is to prepare the second guide and explain to users that we know that they are not the same and that we are working on the facetting.
```suggestion
Multiple filters(_expression_) may be applied to take a broad range of documents and refine them into a more narrow selection, allowing the end user to retrieve the most relevant search results based on the criteria they've selected.

Filters are processed at runtime, contrary to facetted search<Badge text="soon" type="warn"/>. Which gives filters great flexibility but will impact performance.
```

Author: It's maybe too early... Let me think of something.

**MeiliSearch supports filtering through a simple query language.**

### The Query Language

In itself the query language is very simple, and allows you to filter results on any document field. For now it only allows you to query on fields that are either `number`, `boolean` or `string`

### Conditions

> Conditions are the primitives of query filters. They are triplets `field OP value` where:

Reviewer: ```suggestion
Conditions are the primitives of query filters. They are composed of three mandatory parameters: `field OP value` where:
```

> - `field` refers to the document attribute to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `movie_title` are both valid `field`.

Reviewer: ```suggestion
- `field` refers to the document **attribute** to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `movie_title` are both valid `field`.
```

> - `field` refers to the document attribute to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `movie_title` are both valid `field`.

Reviewer: Does this means that if I have an attribute composed of multiple separated words they can be represented without quotes by adding a `_` between each word? Just to be sure I understood correctly

Author: no, the id must be the same as the one in the document. But if you want to filter on a key that's multiple words you ought to use quotes. Any suggestion on how to make it clearer?

Reviewer: By using two different attributes we don't think it's two different way to format the same information. But rather two information formatted differently that works the way you display it.
```suggestion
- `field` refers to the document attribute to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `release_date` are both valid `field`.
```

> - `OP` is the comparison operator, it can be one of `=`, `!=`, , `>`, `>=`, `<` or `<=`.

Reviewer: Empry space between coma's
```suggestion
- `OP` is the comparison **operator**, it can be one of `=`, `!=`, `>`, `>=`, `<` or `<=`.
```

- `value` is the test condition for which the filter shall filter upon.

The `=` and `!=` operators check for equality and difference. For strings, they are both **case insensitive**:

```rust
title = "american ninja" // matches "American Ninja"
```

::: note
`string` values are either double-quoted, single-quoted or a single unquoted word: `title = "Scream"`, `title = 'Scream'` and `title = Scream` are all valid syntaxes, `title = The Avengers` is not.
:::

> The `>`, `>=`, `<` and `<=` apply only to numerical values, and behave the standard way.

Reviewer: ### 
```suggestion
The `>`, `>=`, `<` and `<=` apply **only to numerical values**, and behave the standard way.
```

::: warning
As no specific schema is enforced at indexing, the filtering engine will try to coerce the type of `value`. This can lead to undefined behaviour when big floats are coerced into integers and reciprocally. For this reason, it is best to have homogeneous typing across fields, especially if numbers tend to become large.
:::

### Logical Connectives

An arbitrary number of expressions can be connected together thanks to logical connectives. These connectives are:

- `NOT` negates the following expression, _e.g._ `NOT title = Scream`.
- `OR` performs a logical 'or' between two expressions, _e.g._ `title = Scream OR title = "the avengers"`
- `AND` performs a logical 'and' between two expressions, _e.g._ `title = Dumbo AND title = "Tim Burton"`

`OR` and `AND` are left associative, meaning that the left-hand side is always evaluated first. This is particularly important for `AND`, as if the left-hand side returned `false`, the right-hand side will not be evaluated.

> The simplest form of expressions are conditions. New expressions are created either by connecting other expressions with logical connectives, or by grouping expressions with parentheses, _e.g_:

Reviewer: Just to separate the information of the chapter just above with this one. If you don't think this is useful ignore this suggestion!
```suggestion

### Wording

The simplest form of expressions are conditions. New expressions are created either by connecting other expressions with logical connectives, or by grouping expressions with parentheses, _e.g_:
```

Author: Uhmm, maybe we should make a section called "Expressions" prior to the "logical connectives" section?

Reviewer: Yes, I like your suggestion better. It is better to introduce the wording before digging into it

	},
	{
		"id":"329996",
		"title":"Dumbo",
		"director": "Tim Burton",
		"poster":"https://image.tmdb.org/t/p/w1280/279PwJAcelI4VuBtdzrZASqDPQr.jpg",
		"overview":"A young elephant, whose oversized ears enable him to fly, helps
		save a struggling circus, but when the circus plans a new venture, Dumbo
		and his friends discover dark secrets beneath its shiny veneer.",
		"release_date":1553644800
	},
	{
		"id":"299536",
		"title":"Avengers: Infinity War",
		"director": "Joe Russo",
		"poster":"https://image.tmdb.org/t/p/w1280/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
		"overview":"As the Avengers and their allies have continued to protect the
		world from threats too large for any one hero to handle...",
		"release_date":1524618000
	},
	{
		"id":"458723",
		"title":"Us",
		"director": "Jordan Peele",
		"poster":"https://image.tmdb.org/t/p/w1280/ux2dU1jQ2ACIMShzB3yP93Udpzc.jpg",
		"overview":"Husband and wife Gabe and Adelaide Wilson take their kids
		to their beach house expecting to unplug and unwind with friends. But
		as night descends, their serenity turns to tension and chaos when some
		shocking visitors arrive uninvited.",
		"release_date":1552521600
	},
		...
]
```

> You want to only show to your user movies that were released after a certain date, then you can use the following filter:

Reviewer: ```suggestion
Say you want to only show to your user movies that were released after a certain date, then you can use the following filter:
```

```rust
release_date > 795484800 // march 18, 1995
```

Now imagine that we want only the movies released after the 18 of march 1995, and directed by either Jordan Peel or Tim Burton, then you would use this filter:

```rust
release_date > 795484800 AND (director = "Tim Burton" OR director = "Jordan Peel")
```

Note that filtering on string is case insensitive. Here, the parentheses are mandatory, as `AND` has a higher precedence.

Finally, imagine that you want to filter on "id". You would probably do this:

```rust
id = "299536"
```

> However, this would be a very bad idea. Id uniquely identifies a movie, therefore, no matter how many candidates there are to the user query, only one will ever match. The filtering engine will have to go through all the candidate documents to find the only match. This is highly inefficient and should be avoided. We are working on a [faceted search](https://en.wikipedia.org/wiki/Faceted_search) feature, especialy optimized for this kind of need. It is important to understand that filtering and querying don't have the same performances.

Reviewer: ```suggestion
However, this would be a **very bad idea**. `Id` uniquely identifies a movie, therefore, no matter how many candidates there are to the user query, only one will ever match. The filtering engine will have to go through all the candidate documents to find the only match. This is highly inefficient and should be avoided. We are working on a [faceted search](https://en.wikipedia.org/wiki/Faceted_search) feature, especialy optimized for this kind of need. It is important to understand that filtering and querying don't have the same performances.
```

> However, this would be a very bad idea. Id uniquely identifies a movie, therefore, no matter how many candidates there are to the user query, only one will ever match. The filtering engine will have to go through all the candidate documents to find the only match. This is highly inefficient and should be avoided. We are working on a [faceted search](https://en.wikipedia.org/wiki/Faceted_search) feature, especialy optimized for this kind of need. It is important to understand that filtering and querying don't have the same performances.

Reviewer 2: ```suggestion
However, this would be a very bad idea. `id` uniquely identifies a movie. Therefore, only one document can match the user query. The filtering engine would have to search through all candidate documents to find the only possible match. This would be highly inefficient and should be avoided. We are currently working on a [faceted search](https://en.wikipedia.org/wiki/Faceted_search) feature, especially optimized for this kind of need. It is important to understand that filtering and querying don't have the same performances.
```

# Filtering

A search filter is a parameter that can be applied to the search query to refine the search results of a particular field (e.g. by size, color, price, or brand).
Multiple filters may be applied to take a broad range of documents and refine them into a more narrow selection, allowing the end user to retrieve the most relevant search results based on the criteria they've selected.
**MeiliSearch supports filtering through a simple query language.**

### The Query Language

In itself the query language is very simple, and allows you to filter results on any document field. For now it only allows you to query on fields that are either `number`, `boolean` or `string`

### Conditions

> Conditions are the primitives of query filters. They are triplets `field_id OP value` where:
> 
> - `field_id` refers to the document attribute to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `movie_title` are both valid `field_id`.

Reviewer 3: ```suggestion
Conditions are the primitives of query filters. They are triplets `field OP value` where:

- `field` refers to the document attribute to filter on. (_e.g_ id, title...). It is either a single alphanumeric word or a quoted string: `"movie title"`, and `movie_title` are both valid.
```

> - `OP` is the comparison operator, it can be one of `=`, `!=`, `<`, `>`, `>=` or `<=`.

Reviewer 3: ```suggestion
- `OP` is the comparison operator, it can be one of `=`, `!=`, `>`, `>=`, `<`, or `<=`.
```

- `value` is the test condition for which the filter shall filter upon.

The `=` and `!=` operators check for equality and difference. For strings, they are both **case insensitive**:

```rust
title = "american ninja" // matches "American Ninja"
```

::: note
`string` values are either double-quoted, single-quoted or a single unquoted word: `title = "Scream"`, `title = 'Scream'` and `title = Scream` are all valid syntaxes, `title = The Avengers` is not.
:::

> The `<`, `>`, `>=` and `<=` apply only to numerical values, and behave the standard way.

Reviewer 3: ```suggestion
The `>`, `>=`, `<`, and `<=` apply only to numerical values, and behave the standard way.
```

::: warning
As no specific schema is enforced at indexing, the filtering engine will try to coerce the type of `value`. This can lead to undefined behaviour when big floats are coerced into integers and reciprocally. For this reason, it is best to have homogeneous typing across fields, especially if numbers tend to become large.
:::

### Logical Connectives

An arbitrary number of expressions can be connected together thanks to logical connectives. These connectives are:

> - `NOT`: negates the following expression, _e.g_ `NOT title = Scream`.
> - `OR`: performs a logical 'or' between two expressions, _e.g_ `title = Scream OR title = "the avengers"`
> - `AND`: performs a logical 'and' between two expressions, _e.g_ `title = Scream AND title = "the avengers"`

Reviewer 3: ```suggestion
- `NOT` negates the following expression, _e.g._ `NOT title = Scream`.
- `OR` performs a logical 'or' between two expressions, _e.g._ `title = Scream OR title = "the avengers"`.
- `AND` performs a logical 'and' between two expressions, _e.g._ `title = Dumbo AND director = "Tim Burton"`.
```

> `OR` and `AND` are left associative, meaning that the left-hand side is always evaluated first. This is particularly important for `AND`, as if the left-hand side returned `false`, the right-hand side will not be evaluated.

Reviewer 3: Wouldn't it be better to move that to the "note on performance" part?

Author: It is more than just about performance, it explains behavior, so I think  it belongs to that section, don't you think?

Reviewer 3: I don't think that a user care about the fact that the left or the right branch is evaluated first or second, the result will be the same, the only difference we be in terms of performances.

The simplest form of expressions are conditions. New expressions are created either by connecting other expressions with logical connectives, or by grouping expressions with parentheses, _e.g_:

```rust
title = Scream // a condition is an expression
title = Scream OR title = "The Avengers" // also an expression
(title = Scream OR title = "The Avengers") // use parentheses to isolate an expression
NOT (title = Scream OR title = "The Avengers") // negate the whole expression
rating >= 3 AND (NOT (title = Scream OR title = "The Avengers")) // and so on...
```

::: note
`NOT` has the highest precedence, this means that `NOT title = Scream OR title = "The Avengers"` is effectively evaluated `(NOT title = Scream) OR title = "The Avengers"`. `AND` precedence is lower than `NOT` and higher than `OR`.
:::

> ### A note on performance

Reviewer 3: ```suggestion
### A Note on Performance
```

MeiliSearch is intended to be a flexible tool. For this reason, we decided not to put any restrictions on which fields the user can filter on, nor on how the queries are built. It is important to understand that it can lead to performance issues in some cases. We are currently working on new features to adress this issue, such as [faceting](https://en.wikipedia.org/wiki/Faceted_search) to allow you to efficiently narrow down the number of candidates to a query, and thus improving performances of plain filters, even in worst cases.

### Examples

Suppose that you have a collection of movies, in the following JSON format:

```json
[
	{

> 		"id":"495925",
> 		"title":"Doraemon the Movie: Nobita's Treasure Island",
> 		"director": "Fujiko Fujio",
> 		"poster":"https://image.tmdb.org/t/p/w1280/cmJ71gdZxCqkMUvGwWgSg3MK7pC.jpg",
> 		"overview":"The story is based on Robert Louis Stevenson's Treasure Island novel."
> 		,"release_date":1520035200
> 	},
> 	{
> 		"id":"329996",
> 		"title":"Dumbo",
> 		"director": "Tim Burton",
> 		"poster":"https://image.tmdb.org/t/p/w1280/279PwJAcelI4VuBtdzrZASqDPQr.jpg",
> 		"overview":"A young elephant, whose oversized ears enable him to fly, helps
> 		save a struggling circus, but when the circus plans a new venture, Dumbo
> 		and his friends discover dark secrets beneath its shiny veneer.",
> 		"release_date":1553644800
> 	},
> 	{
> 		"id":"299536",
> 		"title":"Avengers: Infinity War",
> 		"director": "Joe Russo",
> 		"poster":"https://image.tmdb.org/t/p/w1280/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
> 		"overview":"As the Avengers and their allies have continued to protect the
> 		world from threats too large for any one hero to handle...",
> 		"release_date":1524618000
> 	},
> 	{
> 		"id":"458723",
> 		"title":"Us",
> 		"director": "Jordan Peele",
> 		"poster":"https://image.tmdb.org/t/p/w1280/ux2dU1jQ2ACIMShzB3yP93Udpzc.jpg",
> 		"overview":"Husband and wife Gabe and Adelaide Wilson take their kids
> 		to their beach house expecting to unplug and unwind with friends. But
> 		as night descends, their serenity turns to tension and chaos when some
> 		shocking visitors arrive uninvited.",
> 		"release_date":1552521600
> 	}
> 	,

Reviewer 3: ```suggestion
		"id": "495925",
		"title": "Doraemon the Movie: Nobita's Treasure Island",
		"director": "Fujiko Fujio",
		"poster": "https://image.tmdb.org/t/p/w1280/cmJ71gdZxCqkMUvGwWgSg3MK7pC.jpg",
		"overview": "The story is based on Robert Louis Stevenson's Treasure Island novel.",
		"release_date": 1520035200
	},
	{
		"id": "329996",
		"title": "Dumbo",
		"director": "Tim Burton",
		"poster": "https://image.tmdb.org/t/p/w1280/279PwJAcelI4VuBtdzrZASqDPQr.jpg",
		"overview": "A young elephant, whose oversized ears enable him to fly, helps
		save a struggling circus, but when the circus plans a new venture, Dumbo
		and his friends discover dark secrets beneath its shiny veneer.",
		"release_date": 1553644800
	},
	{
		"id": "299536",
		"title": "Avengers: Infinity War",
		"director": "Joe Russo",
		"poster": "https://image.tmdb.org/t/p/w1280/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
		"overview": "As the Avengers and their allies have continued to protect the
		world from threats too large for any one hero to handle...",
		"release_date": 1524618000
	},
	{
		"id": "458723",
		"title": "Us",
		"director": "Jordan Peele",
		"poster": "https://image.tmdb.org/t/p/w1280/ux2dU1jQ2ACIMShzB3yP93Udpzc.jpg",
		"overview": "Husband and wife Gabe and Adelaide Wilson take their kids
		to their beach house expecting to unplug and unwind with friends. But
		as night descends, their serenity turns to tension and chaos when some
		shocking visitors arrive uninvited.",
		"release_date": 1552521600
	},
```

		...
]
```

You want to only show to your user movies that were released after a certain date, then you can use the following filter:

```
release_date > 795484800 // march 18, 1995
```

> Now imagine that we want only the movies released avec the 18 of march 1995, and directed by either Jordan Peel or Tim Burton, then you would use this filter:

Reviewer 3: ```suggestion
Now imagine that we want only the movies released after the 18 of march 1995, and directed by either Jordan Peel or Tim Burton, then you would use this filter:
```

```

> release_date > 795484800 // march 18, 1995 AND (director = "Tim Burton" OR director = "Jordan Peel")

Reviewer 3: ```suggestion
release_date > 795484800 AND (director = "Tim Burton" OR director = "Jordan Peel")
```

# Filtering

> MeiliSearch supports filtering through a simple query language. It allows you

Reviewer: ```suggestion

A search filter is a parameter that can be applied to the search query to refine the search results of a particular field (e.g. by size, color, price, or brand). 

Multiple filters may be applied to take a broad range of documents and refine them into a more narrow selection, allowing the end user to retrieve the most relevant search results based on the criteria they’ve selected.

**MeiliSearch supports filtering through a simple query language.** 
```

> to select the documents that will sent by the server.

Reviewer: ```suggestion
```

### The Query Language

In itself the query language is very simple, and allows you to filter results
on any document field. For now it only allows you to query on fields that are

> either `number`, `bool` or `string`

Reviewer: No reason to make it shorter
```suggestion
either `number`, `boolean` or `string`
```

rating >= 3 AND (NOT (title = Scream OR title = "The Avengers")) // and so on...
```

::: note 
`NOT` has the highest precedence, this means that `NOT title = Scream OR
title = "The Avengers"` is effectively evaluated `(NOT title = Scream) OR
title = "The Avengers"`. `AND` precedence is lower than `NOT` and higher than
`OR`.
:::

### A note on performance

MeiliSearch is intended to be a flexible tool. For this reason, we decided not
to put restrictions on which fields the user can filter on, nor on how the
queries are built. It is important to understand that it can lead to
performance issues in some cases. Thanks to the [bucket
sort](/guides/advanced_guides/bucket_sort), no more documents are filtered than
strictly required to make a response.  Therefore, filtering is very cheap as
there are many documents that match the filter. It becomes an issue when very
few or no document match the filters and there are  many candidates to filter
through. We are working on strategy to address this issue. However, when
performance is critical, and if possible, [faceting](nowhereyet) allows you to very
efficiently narrow down the number of candidates for a query, and thus improving

> performances of plain filters, even in worst cases.

Reviewer 3: Would it be possible to lighten this paragraph?

Reviewer 3: And also to link the faceting to the wikipedia page for the moment?

### Examples

Suppose that you have a collection of books, in the following JSON format:

```json
[
	{
		"id": 0,
		"ISBN-13": "978-0-465-05065-9",
		"title": "The design of everyday things",
		"publisher": "Basic Book",
		"author": "Don Norman",
		"publishing_year": 1988,
		"description": "Even the smartest among us can feel inept as we
		fail to figure out which light switch..."
	},
	{
		"id": 1,
		"ISBN-13": "978-1-593-27828-1",
		"title": "The Rust Programming Language",
		"publisher": "No Starch Press",
		"author": "Steve Klabnik & Carol Nichols",
		"publishing_year": 2019,
		"description": "The official book on the Rust programming
		language, written by the Rust development team at the Mozilla
		Foundation, fully updated for Rust 2018..."
	},
	...
]

> ```

Reviewer 3: Would it be possible to use the movie dataset for a lack of consistency?

Now you want your user to be able select a publisher and get responses for books
only from that publisher. Simply add a filter to the query specifying the
publisher like so:

```
publisher = "no starch press"
```

Note that the filtering on strings is case insensitive. 

Consider the excerpt of our document store above. When the user types "the",
only "The Rust Programming Language" will come out.

Now imagine that we want only the books from "Basic Book", published before
1990, then we can use this filter: 

```
publisher = "Basic Book" AND publishing_year < 1990
```

Finally, imagine that you want to filter on "ISBN-13". You would probably do this:

```
ISBN-13 = "978-0-465-05065-9"
```

However, this would be a very bad idea. ISBN-13 is a number that uniquely
identifies a book, therefore, no matter how many candidates there are to the
user query, only one will ever match this ISBN, and the filtering engine will have
to go through all the candidate documents to find the only match. This is
highly inefficient and should be avoided. You should instead consider using
[faceting](nowhereyet). It is important to understand that filtering and

> querying do not have the same performances.

Reviewer 3: Same here, could you change the link to point to wikipedia?
